cloudant-http/src/main/java/com/cloudant/http/internal/interceptors/CookieInterceptorBase.java

```diff
@@ -0,0 +1,294 @@
+    protected   URL sessionURL;
+
+
+
+    @Override
+    public HttpConnectionInterceptorContext interceptRequest(HttpConnectionInterceptorContext
+                                                                     context) {
+
+        HttpURLConnection connection = context.connection.getConnection();
+
+        // First time we will have no cookies
+        if (cookieManager.getCookieStore().getCookies().isEmpty() && shouldAttemptCookieRequest
+                .get()) {
+            if (!requestCookie(context)) {
+                // Requesting a cookie failed, set a flag if we failed so we won't try again
+                shouldAttemptCookieRequest.set(false);
+            }
+        }
+
+        if (shouldAttemptCookieRequest.get()) {
+
+            // Debug logging
+            if (logger.isLoggable(Level.FINEST)) {
+                logger.finest("Attempt to add cookie to request.");
+                logger.finest("Cookies are stored for URIs: " + cookieManager.getCookieStore()
+                        .getURIs());
+            }
+
+            // Apply any saved cookies to the request
+            try {
+                Map<String, List<String>> requestCookieHeaders = cookieManager.get(connection
+                        .getURL().toURI(), connection.getRequestProperties());
+                for (Map.Entry<String, List<String>> requestCookieHeader :
+                        requestCookieHeaders.entrySet()) {
+                    List<String> cookies = requestCookieHeader.getValue();
+                    if (cookies != null && !cookies.isEmpty()) {
+                        connection.setRequestProperty(requestCookieHeader.getKey(),
+                                listToSemicolonSeparatedString(cookies));
+                    } else {
+                        logger.finest("No cookie values to set.");
+                    }
+                }
+            } catch (IOException e) {
+                logger.log(Level.SEVERE, "Failed to read request properties", e);
+            } catch (URISyntaxException e) {
+                logger.log(Level.SEVERE, "Failed to convert request URL to URI for cookie " +
+                        "retrieval.");
+            }
+        }
+        return context;
+    }
+
+    @Override
+    public HttpConnectionInterceptorContext interceptResponse(HttpConnectionInterceptorContext
+                                                                      context) {
+
+        // Check if this interceptor is valid before attempting any kind of renewal
+        if (shouldAttemptCookieRequest.get()) {
+
+            HttpURLConnection connection = context.connection.getConnection();
+
+            // If we got a 401 or 403 we might need to renew the cookie
+            try {
+                boolean renewCookie = false;
+                int statusCode = connection.getResponseCode();
+
+                if (statusCode == HttpURLConnection.HTTP_FORBIDDEN || statusCode ==
```

Comment:
I think it may be worth factoring parts of this out. The `403` `credentials_expired` checks are very specific to Cloudant I think and it is probably not worth doing all the JSON extraction etc for `403` in IAM cases.

As there is still a decent amount of common functionality that makes it frustrating to override an entire `checkResponse` method we can probably break this out into something like `getRenewalStatusCodes` (which can default to 401) and `performAdditionalChecks()` which can default to a no-op or whatever fits.